generalize circleci docker build.sh and add centos support

jeffdaily opened the pull request:
Add centos Dockerfile and support to circleci docker builds, and allow generic image names to be parsed by build.sh, so both hardcoded images and custom images can be built.  

Currently only adds a ROCm centos Dockerfile.

CC @ezyang @xw285cornell @sunway513

jeffdaily committed:
Add centos support to circleci docker build.

This commit also allows generic image names to be parsed by build.sh
so either hardcoded images or custom images can be built.

Currently only supports a ROCm centos image.

seemethere commented on .circleci/docker/build.sh:
```diff
@@ -247,6 +270,47 @@ case "$image" in
     VISION=yes
     ROCM_VERSION=3.5.1
     ;;
+  *)
```
I feel like this is rather a treatment of a symptom rather than an actual solution.

Like the idea of basing the properties of an image off of the name of said image always seemed like a weird concept to me.

Rather, we should just be explicitly passing the args directly through CircleCI environment variables rather than depending on the name to be within a specific format.

That being said this solution seems okay but brittle in that it relies on magic strings.

Curious to hear your thoughts on this as well @malfet

malfet replied:
I agree with @seemethere that original design is rather bad, but change seems ok if it can be used to pattern-match number of existing cases (which can be deleted as result).
Also, pattern:
```
  if [[ "$image" == *gcc* ]]; then
      GCC_VERSION="$(echo "${image}" | perl -n -e'/gcc(\d+(\.\d+)?)/ && print $1')"
      check_variable GCC_VERSION
    fi
```
can be refactored into something like
```
extract_version_from_image_name gcc GCC_VERSION
```

ezyang replied:
fwiw, the code that currently runs currently does this hybrid "some of the docker properties are based off of naming, but other parts are just hardcoded", since we have a big switch statement on image names which toggle properties. So I think the code here is actually closer to what @seemethere  described as a better state than you might think. There is a sort of chesterton style reason for this, which is that you need to get someone to run the terraform to add a new image name, so people try to avoid doing that as much as possible. 

And weighing in as an old guard, IDM either way you guys go. Just explaining some of the third rails.

jeffdaily replied:
I've added functions to extract all fields from the image name, but it's not currently used anywhere.  Please review latest commit(s) whether we want to use it.

seemethere commented on .circleci/docker/common/install_base.sh:
```diff
@@ -2,52 +2,115 @@
+}
+
+# Install base packages depending on the base OS
+if [ -f /etc/lsb-release ]; then
```
this is a brittle way to do this as well since there is a chance (albeit small) that centos can also contain an `lsb-release` file.

I'd rather base this is off of the `$ID` variable in `/etc/os-release` since that's more consistent, that'll also allow you to use a `case statement here as well instead of a bunch of `if` statements.

jeffdaily replied:
Resolved by 68d0c0e98d967c910b64aacfe9b46042cad3e741.  Please review.

seemethere commented on .circleci/docker/build.sh:
```diff
@@ -29,10 +20,42 @@ elif [[ "$image" == *-bionic* ]]; then
 TRAVIS_DL_URL_PREFIX="https://s3.amazonaws.com/travis-python-archives/binaries/ubuntu/14.04/x86_64"
 
+# verify the given variable was parsed from $image correctly
+function check_variable() {
+  if [ "x$1" = x ]; then
+    echo "missing argument to function check_variable()"
+    exit 1
+  fi
+  if [ "x${!1}" = x ]; then
+    echo "variable '$1' not correctly parsed from image='$image'"
+    exit 1
+  fi
+}
```
Does this function just check if the variable is not blank?

Wouldn't it be easier to just use the built in comparison flags like `-z`?

```suggestion
function check_variable() {
  if [[ -z "$1" ]]; then
    echo "missing argument to function check_variable()"
    exit 1
  fi
  if [[ -z "${!1}" ]]; then
    echo "variable '$1' not correctly parsed from image='$image'"
    exit 1
  fi
}
```

jeffdaily replied:
The intent of `check_variable()` was as much to verify parsing was correct, but also to protect future development of the script in case new variables needed parsing but the caller messed up the syntax.

jeffdaily replied:
Function was removed from latest commit.

malfet commented on .circleci/docker/build.sh:
```diff
@@ -29,10 +20,42 @@ elif [[ "$image" == *-bionic* ]]; then
 elif [[ "$image" == *-focal* ]]; then
   UBUNTU_VERSION=20.04
+elif [[ "$image" == *ubuntu* ]]; then
+  UBUNTU_VERSION="$(echo "${image}" | perl -n -e'/ubuntu(\d+\.\d+)/ && print $1')"
```
Can `$(echo "${image}" | perl -n -e'/${pattern}(\d+\.\d+)/ && print $1')` be refactored into a function?

Because we are doing the same version extraction for `ubuntu`, `py`, `cuda`,`cudnn`,`gcc`,`clang`,rocm`, etc...

jeffdaily replied:
Done.

malfet reviewed (commented):
Can you please rebase your change to get rid of the test failures.

jeffdaily commented:
@malfet @seemethere I'm in the process of addressing your comments.  One general comment, the .circleci/docker/build.sh script is generally useful outside of any automated circleci image builds.  Allowing it to be more general in this manner, being able to parse arbitrary strings as well as adding centos support, helps many of our developers create stable, well-defined images.  It makes sense to us to maintain such scripts as part of the pytorch repo for everyone to take advantage of.

jeffdaily committed:
add functions extract_all_from_image_name, extract_version_from_image_name

jeffdaily commented:
I have not addressed all review comments yet, specifically the `/etc/lsb-release` concern.  But I wanted to see what everyone thought about the two functions I've added to help reduce copy/paste coding, and whether we want to go any further and actually start using the `extract_all_from_image_name` function.

ezyang commented:
Explicitly taking myself off the reviewers. @jeffdaily lmk if you need help unblocking.

jeffdaily committed:
Merge branch 'master' into circleci_centos_image

jeffdaily committed:
update to a more stable OS query

Replaces query of /etc/lsb-release with /etc/os-release.

jeffdaily commented:
@malfet or @seemethere, please comment whether any additional changes are needed, specifically whether any of the hard-coded settings should instead use either of the two new functions `extract_all_from_image_name` or `extract_version_from_image_name`.

jeffdaily committed:
Merge branch 'master' into circleci_centos_image

jeffdaily commented:
Had to rebase again due to #38796 / #42472 .  To avoid more church, please review.  Thanks.

jeffdaily committed:
remove hyphen from rocm and cuda Dockerfile selection

malfet commented on .circleci/docker/common/install_vision.sh:
```diff
@@ -47,11 +47,16 @@ install_centos() {
-  echo "Unable to determine OS..."
-  exit 1
-fi
+ID=$(grep -oP '(?<=^ID=).+' /etc/os-release | tr -d '"')
+case "$ID" in
+  ubuntu)
+    install_ubuntu
+    ;;
+  centos)
+    install_centos
+    ;;
+  *)
+    echo "Unable to determine OS..."
+    exit 1
+    ;;
+esac
```
Hmm, looks like we have an identical tail in `install_vision.sh`, `install_protobuf.sh`, etc, perhaps this could be refactored into a common function?

jeffdaily replied:
Yes, I didn't like the copy/paste coding here.  But I didn't see any mechanism such as a commonly included or sourced script that could be shared between all these `common/install_foo.sh` scripts, because we use docker `ADD common/install_foo.sh` and this would require adding additional files, etc.
